### ISSUE TITLE
Streaming not compatible with logging: 'per-route'

### ISSUE THREAD

**Issue by mateogianolio** [sst/sst#3625]
Using NextjsSite with `streaming: true` and `logging: 'per-route'` breaks logging.

The problem is here: https://github.com/sst/sst/blob/master/packages/sst/src/constructs/SsrFunction.ts#L424

The function passed to `streamifyResponse`:

```ts
awslambda.streamifyResponse(async (event, context) => { /* ... */ }
```

... has the wrong arguments (see https://docs.aws.amazon.com/lambda/latest/dg/configuration-response-streaming.html).

Correct code:

```ts
awslambda.streamifyResponse(async (event, responseStream, context) => {
  /* ... */
  return rawHandler(event, responseStream);
}
```

~They also recommend closing the response stream using `responseStream.end()` or by using `pipe/pipeline` but that's unrelated to this issue. `open-next` does this correctly but it seems sst does not use the open-next server handler?~ **Never mind, I guess `rawHandler` is the openNext server handler?**

---

If you've already posted your issue on Discord, make sure to leave a link to it here.
https://discord.com/channels/983865673656705025/1083422760111439973/threads/1196815754230771712

**Comment by mateogianolio** [comment]
Ah this problem also causes the warmer function to time out if using `streaming: true`. Streaming responses ignore return types and must be explicitly closed with `responseStream.end()`:

```
if (event.type === "warmer") {
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve({ serverId: "server-" + Math.random().toString(36).slice(2, 8) });
    }, event.delay);
  });
}
```

**Comment by mateogianolio** [comment]
Here's the patch I applied for sst@2.39.6:

```
diff --git a/node_modules/sst/constructs/SsrFunction.js b/node_modules/sst/constructs/SsrFunction.js
index 5d8cbb1..ebbd05a 100644
--- a/node_modules/sst/constructs/SsrFunction.js
+++ b/node_modules/sst/constructs/SsrFunction.js
@@ -274,10 +274,10 @@ export class SsrFunction extends Construct {
         const newHandlerFunction = "handler";
         await fs.writeFile(path.join(bundle, handlerDir, `${newHandlerName}.mjs`), streaming
             ? [
-                `export const ${newHandlerFunction} = awslambda.streamifyResponse(async (event, context) => {`,
+                `export const ${newHandlerFunction} = awslambda.streamifyResponse(async (event, responseStream, context) => {`,
                 ...injections,
                 `  const { ${oldHandlerFunction}: rawHandler} = await import("./${oldHandlerName}.mjs");`,
-                `  return rawHandler(event, context);`,
+                `  return rawHandler(event, responseStream);`,
                 `});`,
             ].join("\n")
             : [
diff --git a/node_modules/sst/constructs/SsrSite.js b/node_modules/sst/constructs/SsrSite.js
index af27a7d..630187a 100644
--- a/node_modules/sst/constructs/SsrSite.js
+++ b/node_modules/sst/constructs/SsrSite.js
@@ -648,11 +648,18 @@ function handler(event) {
         function useServerFunctionWarmingInjection() {
             return `
 if (event.type === "warmer") {
-  return new Promise((resolve) => {
+  if (typeof responseStream !== "undefined") {
     setTimeout(() => {
-      resolve({ serverId: "server-" + Math.random().toString(36).slice(2, 8) });
+      responseStream.write(JSON.stringify({ serverId: "server-" + Math.random().toString(36).slice(2, 8) }));
+      responseStream.end();
     }, event.delay);
-  });
+  } else {
+    return new Promise((resolve) => {
+      setTimeout(() => {
+        resolve({ serverId: "server-" + Math.random().toString(36).slice(2, 8) });
+      }, event.delay);
+    });
+  }
 }`;
         }
         function getS3FileOptions(copy) {

```

**Comment by jayair** [comment]
Oh appreciate this!

**Comment by mateogianolio** [comment]
Note: I cleaned up the patch a bit above and added a missed `return` statement for the streaming response :)

**Comment by fwang** [comment]
Thanks @mateogianolio! Released in v2.39.9.

Btw. does this also fix the logging problem with logging?

**Comment by mateogianolio** [comment]
Yes it does!
